## О Jupyter notebook

Jupyter Notebook — это среда разработки, где сразу можно видеть результат выполнения кода и его отдельных фрагментов. Отличие от традиционной среды разработки в том, что код можно разбить на куски и выполнять их в произвольном порядке.
Для запуска кода в ячейке можно воспользоваться меню сверху или нажать клавиши "Enter+Shift".

Основная область применения ноутбуков — машинное обучение, нейросети, визуализация данных и статистика.

Ещё такую среду часто используют для поэтапной разработки, когда нужно по шагам проверять работу разных фрагментов кода. Дело в том, что код в ноутбуках хранится в независимых ячейках и его можно запускать в любом порядке или поодиночке. Это позволяет быстро экспериментировать с алгоритмами и находить оптимальное решение.

Также jupyter-ноутбуки используются при обучении программированию на Python, чтобы писать код и сразу видеть результат его работы.

Есть два основных типа ячеек, которые мы рассмотрим:

- Ячейка кода содержит код, который должен быть выполнен в ядре, и отображает его вывод ниже.
- Ячейка Markdown содержит текст, отформатированный с использованием Markdown, и отображает его вывод на месте при запуске.

In [1]:
# ячейка с кодом, при выполнении которой появится output
2 + 2

4

А это ___ячейка с текстом___.

Попробуйте создать свои ячейки, написать какой-нибудь код и текст какой-нибудь формулой.

In [2]:
# your code
4 + 45

49

[Здесь](https://athena.brynmawr.edu/jupyter/hub/dblank/public/Jupyter%20Notebook%20Users%20Manual.ipynb) находится <s>не</s>большая заметка о используемом языке разметки Markdown. Он позволяет:

0. Составлять упорядоченные списки
1. #Делать 
##заголовки 
###разного уровня
3. Выделять *текст* <s>при</s> **необходимости**
4. Добавлять [ссылки](http://imgs.xkcd.com/comics/the_universal_label.png)


* Составлять неупорядоченные списки

Делать вставки с помощью LaTex:
    
$
\left\{
\begin{array}{ll}
x = 16 \sin^3 (t) \\ 
y = 13 \cos (t) - 5 \cos (2t) - 2 \cos (3t) - \cos (4t) \\
t \in [0, 2 \pi]
\end{array}
\right.$

## 1. Табличные данные и Pandas

Pandas — это библиотека для работы с данными на Python. Она упрощает жизнь аналитикам: где раньше использовалось 10 строк кода теперь хватит одной.

Например, чтобы прочитать данные из csv, в стандартном Python надо сначала решить, как хранить данные, затем открыть файл, прочитать его построчно, отделить значения друг от друга и очистить данные от специальных символов.

В Pandas всё проще. Во-первых, не нужно думать, как будут храниться данные — они лежат в датафрейме. Во-вторых, достаточно написать одну команду:

 data = pd.read_csv('file.csv')

Pandas добавляет в Python новые структуры данных — серии и датафреймы. 

Серии — одномерные массивы данных. Они очень похожи на списки, но отличаются по поведению — например, операции применяются к списку целиком, а в сериях — поэлементно.
Еще одно отличие серий от списков — в качестве индексов можно использовать произвольные значения, это делает данные нагляднее.
Датафрейм — это проиндексированный многомерный массив значений, соответственно каждый столбец датафрейма, является серией.

In [3]:
#%pylab inline # import almost all we need
import pandas

#### 1. Прочитайте файл с таблицей, выведите последние 10 строк.



In [51]:
data = pandas.read_csv('data_for_lab-1.csv', delimiter=',')
data.tail(10)

event_time             order_id           product_id  \
2633511  2020-11-21 10:05:04 UTC  2388440981134693938  1515966223509088509   
2633512  2020-11-21 10:06:01 UTC  2388440981134693939  1515966223509090132   
2633513  2020-11-21 10:08:14 UTC  2388440981134693940  2273948246856434093   
2633514  2020-11-21 10:08:54 UTC  2388440981134693941  1515966223519279912   
2633515  2020-11-21 10:08:54 UTC  2388440981134693941  1515966223509088622   
2633516  2020-11-21 10:10:01 UTC  2388440981134693942  1515966223526602848   
2633517  2020-11-21 10:10:13 UTC  2388440981134693943  1515966223509089282   
2633518  2020-11-21 10:10:30 UTC  2388440981134693944  1515966223509089917   
2633519  2020-11-21 10:10:30 UTC  2388440981134693944  2273948184839454837   
2633520  2020-11-21 10:10:30 UTC  2388440981134693944  1515966223509127566   

          category_id                  category_code     brand    price  \
2633511  2.268105e+18         electronics.smartphone     apple   925.67   
2633512  2.268105e+18  computers.peripherals.printer     canon   299.98   
2633513  2.268105e+18                            NaN    xiaomi    16.18   
2633514  2.374499e+18           electronics.video.tv   samsung  1736.09   
2633515  2.374499e+18           electronics.video.tv        lg   370.35   
2633516  2.268105e+18         electronics.smartphone      oppo   138.87   
2633517  2.268105e+18         electronics.smartphone     apple   418.96   
2633518  2.268105e+18     appliances.personal.scales     vitek    12.48   
2633519  2.268105e+18                            NaN  moulinex    41.64   
2633520  2.268105e+18     appliances.kitchen.blender   redmond    53.22   

              user_id  
2633511  1.515916e+18  
2633512  1.515916e+18  
2633513  1.515916e+18  
2633514  1.515916e+18  
2633515  1.515916e+18  
2633516  1.515916e+18  
2633517  1.515916e+18  
2633518  1.515916e+18  
2633519  1.515916e+18  
2633520  1.515916e+18

#### 2.  Выведите следующую информацию:
1. Сколько всего заказов в исходной выборке данных?
2. Количество уникальных категорий товара (category_code)

In [52]:
print('Count:',len(data.order_id.unique()))
print('Count:',len(data['category_code'].unique()))

Count: 1435266
Count: 511


#### 3. Найдите столбцы с пустыми значениями, заполните строковые значения пустой строкой, числовые - 0.

In [53]:
data.category_code.fillna('',inplace=True)
data.isna().sum()
data=data
data

event_time             order_id           product_id  \
0        2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
1        2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
2        2020-04-24 14:37:43 UTC  2294444024058086220  2273948319057183658   
3        2020-04-24 14:37:43 UTC  2294444024058086220  2273948319057183658   
4        2020-04-24 19:16:21 UTC  2294584263154074236  2273948316817424439   
...                          ...                  ...                  ...   
2633516  2020-11-21 10:10:01 UTC  2388440981134693942  1515966223526602848   
2633517  2020-11-21 10:10:13 UTC  2388440981134693943  1515966223509089282   
2633518  2020-11-21 10:10:30 UTC  2388440981134693944  1515966223509089917   
2633519  2020-11-21 10:10:30 UTC  2388440981134693944  2273948184839454837   
2633520  2020-11-21 10:10:30 UTC  2388440981134693944  1515966223509127566   

          category_id                category_code     brand   price  \
0        2.268105e+18           electronics.tablet   samsung  162.01   
1        2.268105e+18           electronics.tablet   samsung  162.01   
2        2.268105e+18  electronics.audio.headphone    huawei   77.52   
3        2.268105e+18  electronics.audio.headphone    huawei   77.52   
4        2.268105e+18                                karcher  217.57   
...               ...                          ...       ...     ...   
2633516  2.268105e+18       electronics.smartphone      oppo  138.87   
2633517  2.268105e+18       electronics.smartphone     apple  418.96   
2633518  2.268105e+18   appliances.personal.scales     vitek   12.48   
2633519  2.268105e+18                               moulinex   41.64   
2633520  2.268105e+18   appliances.kitchen.blender   redmond   53.22   

              user_id  
0        1.515916e+18  
1        1.515916e+18  
2        1.515916e+18  
3        1.515916e+18  
4        1.515916e+18  
...               ...  
2633516  1.515916e+18  
2633517  1.515916e+18  
2633518  1.515916e+18  
2633519  1.515916e+18  
2633520  1.515916e+18  

[2633521 rows x 8 columns]

#### 4.  Обратите внимание на колонку, содержащую цену товара. Создайте новую колонку, таким образом, чтобы цена стала числом.

Для этого необходимо применить функцию-преобразование к каждой строке таблицы (для этого есть соответствующая функция).

Выведите среднюю/минимальную/максимальную цену у товара.

In [54]:
data['price_ch'] = pandas.to_numeric(data['price'])
maxim = data['price_ch'].max()
meane = data['price_ch'].mean()
medan = data['price_ch'].median()
minsk = data['price_ch'].min()
print(maxim)
print(meane)
print(medan)
print(minsk)

50925.9
154.09316532704963
55.53
0.0


Удалите старую колонку с ценой.

In [55]:
data=data.drop(columns = ['price'],axis = 0)
data

event_time             order_id           product_id  \
0        2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
1        2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
2        2020-04-24 14:37:43 UTC  2294444024058086220  2273948319057183658   
3        2020-04-24 14:37:43 UTC  2294444024058086220  2273948319057183658   
4        2020-04-24 19:16:21 UTC  2294584263154074236  2273948316817424439   
...                          ...                  ...                  ...   
2633516  2020-11-21 10:10:01 UTC  2388440981134693942  1515966223526602848   
2633517  2020-11-21 10:10:13 UTC  2388440981134693943  1515966223509089282   
2633518  2020-11-21 10:10:30 UTC  2388440981134693944  1515966223509089917   
2633519  2020-11-21 10:10:30 UTC  2388440981134693944  2273948184839454837   
2633520  2020-11-21 10:10:30 UTC  2388440981134693944  1515966223509127566   

          category_id                category_code     brand       user_id  \
0        2.268105e+18           electronics.tablet   samsung  1.515916e+18   
1        2.268105e+18           electronics.tablet   samsung  1.515916e+18   
2        2.268105e+18  electronics.audio.headphone    huawei  1.515916e+18   
3        2.268105e+18  electronics.audio.headphone    huawei  1.515916e+18   
4        2.268105e+18                                karcher  1.515916e+18   
...               ...                          ...       ...           ...   
2633516  2.268105e+18       electronics.smartphone      oppo  1.515916e+18   
2633517  2.268105e+18       electronics.smartphone     apple  1.515916e+18   
2633518  2.268105e+18   appliances.personal.scales     vitek  1.515916e+18   
2633519  2.268105e+18                               moulinex  1.515916e+18   
2633520  2.268105e+18   appliances.kitchen.blender   redmond  1.515916e+18   

         price_ch  
0          162.01  
1          162.01  
2           77.52  
3           77.52  
4          217.57  
...           ...  
2633516    138.87  
2633517    418.96  
2633518     12.48  
2633519     41.64  
2633520     53.22  

[2633521 rows x 8 columns]

#### 5. Отсортируйте стоимость товаров по убыванию, выведите 7 дорогих товаров. Отсортируйте товары по возрастанию стоимости, выведите 7 дешевых товаров  (по choice_description)

Для этого избавьтесь от дубликатов и отсортируйте товары. Не забудьте про количество товара.

In [56]:
data.drop_duplicates(subset=['product_id'])
(data.sort_values('price_ch', ascending=False)).head(7)

event_time             order_id           product_id  \
1144303  2020-05-26 08:12:38 UTC  2353234364520727464  1515966223509105016   
28182    2020-06-25 10:08:37 UTC  2339244674033647628  1515966223509131884   
28623    2020-06-26 09:10:44 UTC  2339940319191106553  1515966223509566954   
1153319  2020-05-27 06:37:30 UTC  2353235498996073224  2273948305316643078   
1627981  2020-07-03 11:59:01 UTC  2353288509000777918  2273948305316643078   
2270999  2020-09-16 06:46:10 UTC  2388440981134484484  1515966223520801280   
8742     2020-05-14 09:09:41 UTC  2308774433587724743  1515966223509105377   

          category_id           category_code    brand       user_id  price_ch  
1144303  2.374499e+18    electronics.video.tv  samsung           NaN  50925.90  
28182    2.374499e+18    electronics.video.tv      NaN  1.515916e+18  18328.68  
28623    2.374499e+18    electronics.video.tv      NaN  1.515916e+18  13310.16  
1153319  2.268105e+18    electronics.video.tv       lg           NaN  11574.05  
1627981  2.268105e+18    electronics.video.tv       lg  1.515916e+18  11574.05  
2270999  2.374499e+18    electronics.video.tv  samsung  1.515916e+18  10416.64  
8742     2.268105e+18  electronics.smartphone       lg  1.515916e+18   9606.48

In [10]:
data.drop_duplicates(subset=['product_id'])
(data.sort_values('price_ch', ascending=False)).tail(7)

event_time             order_id           product_id  \
877360  2020-03-20 09:07:26 UTC  2348823338635232193  1515966223523303314   
493382  2020-02-15 13:21:13 UTC  2348795511374348846  1515966223523303310   
493380  2020-02-15 13:21:13 UTC  2348795511374348846  1515966223523303366   
493378  2020-02-15 13:21:13 UTC  2348795511374348846  1515966223523303312   
493377  2020-02-15 13:21:13 UTC  2348795511374348846  1515966223523303340   
493376  2020-02-14 13:42:25 UTC  2348795510661317165  1515966223523303302   
524287  2020-02-19 07:07:09 UTC  2348798111096243168  1515966223523303301   

        category_id category_code                brand  price  user_id  \
877360          0.0          9.26  1515915625484674144    0.0      0.0   
493382          0.0          8.10                    0    0.0      0.0   
493380          0.0          4.63                    0    0.0      0.0   
493378          0.0          6.94                    0    0.0      0.0   
493377          0.0          5.79                    0    0.0      0.0   
493376          0.0         57.87                    0    0.0      0.0   
524287          0.0         16.18                    0    0.0      0.0   

        price_ch  
877360       0.0  
493382       0.0  
493380       0.0  
493378       0.0  
493377       0.0  
493376       0.0  
524287       0.0

#### 6. Выведите информацию о том, сколько раз клиенты покупали больше 1 товара apple (brand)?

In [57]:
#data[data['brand']=='apple'].value_counts().to_frame()
df=data[data['brand']=='apple']
#filter1 = df['brand'].isin(['apple'])
#df[filter1]
count = df[df.duplicated()]
print(len(count['order_id'].unique()))

23


#### 7. Выведите информацию о среднем чеке у заказа и сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые значения.

In [61]:
round(data.groupby('order_id')['price_ch'].sum().mean())

236

In [62]:
round(data.groupby('order_id')['price_ch'].count().mean())

2

#### 8. Выведите количество заказов с 1 товаром.

#### 9. Выведите самую популярную категорию товара.

In [63]:
data['category_code'].mode()

0    
dtype: object

#### 10. Выведите виды товаров Apple. Какой из них чаще всего покупают? Какой из них самый дорогой? 

In [64]:
filter1 = data['brand'].isin(['apple'])
df10=data[filter1]
data10=df10.drop_duplicates(subset='category_code')
print(data10['category_code'])
popular_Apple=df10['category_code'].mode()
print(popular_Apple)
expensive=df10['price_ch'].max()
print(expensive)

6               electronics.smartphone
34         electronics.audio.headphone
36                  electronics.clocks
162                                   
303                 computers.notebook
439                 electronics.tablet
687           construction.tools.screw
1283          computers.components.hdd
4737                 computers.desktop
6721    computers.peripherals.keyboard
6725       computers.peripherals.mouse
9544       computers.components.cooler
Name: category_code, dtype: object
0    electronics.smartphone
dtype: object
6018.5


#### 11. В каком количестве заказов есть товар, который стоит более 40% от суммы всего чека?

Возможно, будет удобно посчитать отдельно средний чек, добавить его в исходные данные и сделать необходимые проверки.

In [69]:
tmp = data.groupby('order_id')['price_ch'].sum()*0.4   #не готово
d11 = pandas.merge(data,tmp,on ='order_id')
#len(d11[d11['price']>d11['price_ch']]['order_id'].unique())

#### 12. Предположим, что в данных была ошибка и товар с product_id 1515966223509089906, который стоил 162.01, должен был стоить 170,02. Скорректируйте данные в таблицы и посчитайте, на какой процент больше денег было заработано с этого товара. Не забывайте, что количество товара не всегда равно 1.

In [71]:
filter2 = data['product_id'].isin(['1515966223509089906'])
df12=data[filter2]
sum1=df12['price_ch'].sum()
df12.loc[(df12.price_ch!=170.02), 'price_ch']=170.02
sum2=df12['price_ch'].sum()
result=sum2*100/sum1-100
print(result)
df12

4.9441392506635395


C:\Users\Renat\anaconda3\lib\site-packages\pandas\core\indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


event_time             order_id           product_id  \
0        2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
1        2020-04-24 11:50:39 UTC  2294359932054536986  1515966223509089906   
6123     2020-05-10 12:33:12 UTC  2305977760524272239  1515966223509089906   
9556     2020-05-16 07:58:26 UTC  2310188121796378674  1515966223509089906   
10225    2020-05-17 15:00:08 UTC  2311125144107484149  1515966223509089906   
...                          ...                  ...                  ...   
2604076  2020-11-14 09:36:15 UTC  2388440981134677150  1515966223509089906   
2611374  2020-11-14 11:48:43 UTC  2388440981134680889  1515966223509089906   
2615842  2020-11-18 09:31:13 UTC  2388440981134683904  1515966223509089906   
2619162  2020-11-18 12:18:52 UTC  2388440981134685713  1515966223509089906   
2625836  2020-11-19 09:43:36 UTC  2388440981134689515  1515966223509089906   

          category_id       category_code    brand       user_id  price_ch  
0        2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
1        2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
6123     2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
9556     2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
10225    2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
...               ...                 ...      ...           ...       ...  
2604076  2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
2611374  2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
2615842  2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
2619162  2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  
2625836  2.268105e+18  electronics.tablet  samsung  1.515916e+18    170.02  

[884 rows x 8 columns]

#### 13. Создайте новый DateFrame из матрицы, созданной ниже. Назовите колонки index, column1, column2 и сделайте первую колонку индексом.

In [72]:
import numpy as np
df = np.random.rand(10, 3)

df = pandas.DataFrame(df[:,1:],columns=['Column1','Column2'],index = df[:,0])
df

Column1   Column2
0.786239  0.912361  0.563931
0.053793  0.071472  0.765194
0.258173  0.771474  0.869065
0.258739  0.395341  0.811418
0.346197  0.397667  0.875763
0.874178  0.115337  0.140624
0.352316  0.317570  0.002299
0.288186  0.696783  0.234590
0.908774  0.309165  0.266728
0.241616  0.689895  0.677852

Сохраните DataFrame локально в формате csv без индексов и названий столбцов.

In [19]:
data.to_csv('DM.csv', header=False, index=False)

## 2. Визуализации и matplotlib

Библиотека matplotlib - это бибилиотека двумерной графики для языка программирования python с помощью которой можно создавать высококачественные рисунки различных форматов. Matplotlib представляет собой модуль-пакет для python.
Matplotlib cостоит из множества модулей. Модули наполнены различными классами и функциями, которые иерархически связаны между собой.

In [20]:
%matplotlib inline  # нужно для отображения графиков внутри ноутбука
import matplotlib.pyplot as plt

UsageError: unrecognized arguments: # нужно для отображения графиков внутри ноутбука


На самом деле мы уже импортировали matplotlib внутри %pylab inline в начале задания.

Работать мы будем с той же выборкой покупкок. Добавим новую колонку с датой покупки.

In [ ]:
import datetime
import random

start = datetime.datetime(2018, 1, 1)
end = datetime.datetime(2018, 1, 31)
delta_seconds = int((end - start).total_seconds())

dates = pandas.DataFrame(index=data.order_id.unique())
dates['date'] = [(start + datetime.timedelta(seconds=random.randint(0, delta_seconds))).strftime('%Y-%m-%d')
                 for _ in range(data.order_id.nunique())]

# если DataFrame с покупками из прошлого заказа называется не df, замените на ваше название ниже
data['date'] = data.order_id.map(dates['date'])
data

#### 1. Постройте гистограмму распределения сумм покупок и гистограмму средних цен отдельных видов продуктов product_id. 

Изображайте на двух соседних графиках. Для используйте subplot.

In [ ]:
# your code

#### 2. Постройте график зависимости суммы покупок от дней.

In [ ]:
# your code

#### 3. Постройте средних сумм покупок по дням недели (bar plot).

In [ ]:
# your code

#### 4. Постройте график зависимости денег за товар от купленного количества (scatter plot).

In [ ]:
# your code

Сохраните график в формате pdf (так он останется векторизованным).

In [ ]:
# your code

Кстати, существует надстройка над matplotlib под названием [seaborn](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html). Иногда удобнее и красивее делать визуализации через неё. 